File: 0104-Maximum Depth of Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int maxDepth(TreeNode* root) {
        if(root){
            return dfs(root, 1);
        }
        else{
            return 0;
        }
    }
    
    int dfs(TreeNode* root, int depth){
        if(root->left && root ->right){
            return 1 + max(dfs(root->left, depth), dfs(root->right, depth));
        }
        if(root->left){
            return 1 + dfs(root->left, depth);
        }
        if(root->right){
            return 1 + dfs(root->right, depth);
        }
        return depth;
    }
};

File: 0104-Maximum Depth of Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def maxDepth(self, root: TreeNode) -> int:
        if root:
            depth = 1
            return self.dfs(root, depth)
        else:
            return 0
    
    def dfs(self, root, depth):
        if root.left and root.right:
            depth = max(self.dfs(root.left, depth), self.dfs(root.right, depth)) + 1
            return depth
        if root.left:
            depth = self.dfs(root.left, depth) + 1
            return depth
        if root.right:
            depth = self.dfs(root.right, depth) + 1
            return depth
        return depth